### PR TITLE
Angular 9 / 10 Upgrade

### LINKED ISSUE #1255: Angular 9/10 Upgrade
Upgrade Cashmere to Angular 9.

Long-term support of Angular 8 [ends on Nov 28, 2020](https://angular.io/guide/releases#versioning).  So we should have Cashmere upgraded to version 9 by then.

We want to make sure and give app teams plenty of time to plan for this change.  So well in advance lets be sure and broadcast the coming change.

### PR DESCRIPTION
**Pull request by andrew-frueh**
At long last, I think this is ready for a PR. This was a massive slog, and many thanks to @joeskeen for all his help.  Because this was such a big overhaul, I'm going to wait to merge until at least three of you have had a chance to review - as well as pull the branch locally and validate that you can successfully build.  I know it's a big PR, but this one definitely needs the extra eyes.  I'm happy to kick in Motivosity for you all taking the time.  Also happy to answer any questions as best I can about the changes, only a few of them fall into the 'voodoo' category.

closes #1255

### CODE REVIEW

**Comment by andrew-frueh** [comment]
Also leaving a note-to-self that after we merge, need to validate that StackBlitz isn't broken.  50/50 odds on that.  We might need a quick patch to fix if anything is broken before we announce the release.

**Comment by andrew-frueh** [comment]
Looks like I've got three tests failing on table dataSource.  That's annoying because if you run `npm run test`, the first time it passes, and then the second run it fails.  I figured because of that it wouldn't be an issue on Travis, but apparently it follows the same pattern - first time pass, next time fail.  I'll work fixing those.

File: projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts [hunk 1]
```diff
@@ -7,20 +7,21 @@ describe('AppSwitcherService', () => {
...
-        beforeEach(() => (service = TestBed.get(AppSwitcherService)));
+        beforeEach(() => (service = TestBed.inject(AppSwitcherService)));
 
         it('should have the correct allApplicationsUri', () => {
             expect(service.allApplicationsUri).toBe('foo://bar/baz/apps');
         });
 
-        it('should use the correct uri when fetching services', async(
+        // TEST HAS NO EXPECTATIONS
+        xit('should use the correct uri when fetching services', waitForAsync(
```
Comment (corykon) [inline projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts:24]: should we delete this guy?

File: projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts [hunk 2]
```diff
@@ -37,20 +38,21 @@ describe('AppSwitcherService', () => {
...
 
-        beforeEach(() => (service = TestBed.get(AppSwitcherService)));
+        beforeEach(() => (service = TestBed.inject(AppSwitcherService)));
 
         it('should have the correct allApplicationsUri', () => {
             expect(service.allApplicationsUri).toBe('foo://bar/baz/apps');
         });
 
-        it('should use the correct uri when fetching services', async(
+        // TEST HAS NO EXPECTATIONS
```
Comment (corykon) [inline projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts:54]: delete?

File: src/app/components/component-viewer/component-usage/component-usage.component.ts [hunk 3]
```diff
@@ -18,11 +19,37 @@ export class ComponentUsageComponent {
...
     }
+
+    private removeLines(pre: HTMLPreElement): void {
+        const span = pre.querySelector('span.line-number');
+        if (span) {
+            pre.removeChild(span);
+        }
+    }
+
+    private addLines(pre: HTMLPreElement): void {
```
Comment (corykon) [inline src/app/components/component-viewer/component-usage/component-usage.component.ts:45]: No way to reuse the new highlight directive to avoid repeating code here?

**Review by corykon — APPROVED** [review]
Everything's building, tests passing, local env looks good 👍 
Left a couple comments on super minor things, and pushed one little fix for a menu arrow that was hanging outside its container - 
image

Let's get this thing finished!!

File: projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts (hunk 1 above)
Comment (andrew-frueh) [reply projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts:24]: I figured these were stubbed in to be added at a later date - but who knows?

File: src/app/components/component-viewer/component-usage/component-usage.component.ts (hunk 3 above)
Comment (andrew-frueh) [reply src/app/components/component-viewer/component-usage/component-usage.component.ts:45]: yeah that's probably worth a look - I was getting pretty lazy by the end.  I'll see if there is a way to reuse.

**Review by Techgeekster — APPROVED** [review]

File: src/app/components/component-viewer/component-usage/component-usage.component.ts (hunk 3 above)
Comment (andrew-frueh) [reply src/app/components/component-viewer/component-usage/component-usage.component.ts]: good suggestion @corykon - got that updated so the markdown directive and example viewer both leverage the HighlightDirective.  Was slightly tricky with the markdown parsing because the directive has to get added dynamically, but works just fine.  And now there aren't three duplicates of the same functions 👍

**Comment by andrew-frueh** [comment]
Thanks for the review @Techgeekster!  I know you'll be one of the first to pull this into your app, so the real test will be seeing if anything blows up there that we need to patch.

File: projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts (hunk 1 above)
Comment (joeskeen) [reply projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts:24]: I think the reason it reports there is no expectation is that the mock.verify() happens after the test completes, wrapping with waitForAsync() may fix it

File: projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts (hunk 2 above)
Comment (joeskeen) [reply projects/cashmere/src/lib/app-switcher/app-switcher.service.spec.ts:54]: same as above I think

File: projects/cashmere/src/lib/button/split-button/split-button.component.ts [hunk 4]
```diff
@@ -83,10 +83,7 @@ export class SplitButtonComponent {
         this._tabIndex = value == null ? 0 : value;
     }
 
-    /**
-     * @deprecated
-     * @description Use `buttonStyle` instead
-     * */
+    /** **DEPRECATED** Use `buttonStyle` instead */
```
Comment (joeskeen) [inline projects/cashmere/src/lib/button/split-button/split-button.component.ts:86]: When (if ever) do we plan to remove this deprecated property?  Also, curious why we aren't using the JSDoc comments for this anymore?

File: projects/cashmere/src/lib/button/button.component.ts [hunk 5]
```diff
@@ -38,10 +38,7 @@ export class ButtonComponent {
     private _style: string;
     private _size: string;
 
-    /**
-     * @deprecated
-     * @description Use `buttonStyle` instead
-     * */
+    /** **DEPRECATED** Use `buttonStyle` instead */
```
Comment (joeskeen) [inline projects/cashmere/src/lib/button/button.component.ts:41]: similarly, when (if ever) do we plan to remove this deprecation?

File: projects/cashmere/src/lib/chip/chip.component.ts [hunk 6]
```diff
@@ -67,10 +67,7 @@ export class ChipComponent {
         this.closeClick.emit(e);
     }
 
-    /**
-     * @deprecated
-     * @description Use `hasCloseButton` instead
-     * */
+    /** **DEPRECATED** Use `hasCloseButton` instead */
```
Comment (joeskeen) [inline projects/cashmere/src/lib/chip/chip.component.ts:70]: same here

File: projects/cashmere/src/lib/datepicker/datetime/date-adapter.ts [hunk 7]
```diff
@@ -21,10 +21,8 @@ export function HC_DATE_LOCALE_FACTORY(): string {
 }
 
 /**
- * No longer needed since HC_DATE_LOCALE has been changed to a scoped injectable.
+ * **DEPRECATED** No longer needed since HC_DATE_LOCALE has been changed to a scoped injectable.
```
Comment (joeskeen) [inline projects/cashmere/src/lib/datepicker/datetime/date-adapter.ts:24]: same here

File: projects/cashmere/src/lib/form-field/hc-form-field.component.spec.ts [hunk 8]
```diff
@@ -74,12 +74,12 @@ describe('HcFormFieldComponent', () => {
             expect(labelElement.textContent).toMatch(/^hola$/);
         });
 
-        it('support type attribute', () => {});
+        // it('support type attribute', () => {});
```
Comment (joeskeen) [inline projects/cashmere/src/lib/form-field/hc-form-field.component.spec.ts:77]: We should remove empty commented out code here

File: projects/cashmere/src/lib/sort/sort.spec.ts [hunk 9]
```diff
@@ -42,7 +42,7 @@
 //
 //   let component: SimpleMatSortApp;
 //
-//   beforeEach(async(() => {
+//   beforeEach(waitForAsync(() => {
```
Comment (joeskeen) [inline projects/cashmere/src/lib/sort/sort.spec.ts:45]: do we really need this file?

File: projects/cashmere/src/lib/table/table.spec.ts [hunk 10]
```diff
@@ -26,7 +26,7 @@
 //
 //
 // describe('MatTable', () => {
-//   beforeEach(async(() => {
+//   beforeEach(waitForAsync(() => {
```
Comment (joeskeen) [inline projects/cashmere/src/lib/table/table.spec.ts:29]: is this file needed?

File: projects/cashmere/src/lib/tile/tile-header.directive.spec.ts [hunk 11]
```diff
@@ -46,8 +46,8 @@ describe('TileHeaderDirective', () => {
         expect(directive._headerBlue).toBeFalsy();
     });
 
-    it('when setting an invalid value', () => {
-        const setInvalid = () => component.headerType = 'someInvalidValue';
+    describe('when setting an invalid value', () => {
```
Comment (joeskeen) [inline projects/cashmere/src/lib/tile/tile-header.directive.spec.ts:49]: nice catch!

File: projects/cashmere/src/lib/toaster/hc-toast-ref.ts [hunk 12]
```diff
@@ -2,8 +2,9 @@ import {OverlayRef} from '@angular/cdk/overlay';
 import {HcToastComponent} from './hc-toast.component';
 import {Subject} from 'rxjs';
 import {filter, take} from 'rxjs/operators';
-import {Input} from '@angular/core';
+import { Input, Directive } from '@angular/core';
 
+@Directive()
```
Comment (joeskeen) [inline projects/cashmere/src/lib/toaster/hc-toast-ref.ts:7]: Is this really a Directive?  Seems like it should instead just be an `@Injectable()`

**Comment by joeskeen** [comment]
I give my 👍 once my comments are resolved.

File: projects/cashmere/src/lib/button/split-button/split-button.component.ts (hunk 4 above)
Comment (andrew-frueh) [reply projects/cashmere/src/lib/button/split-button/split-button.component.ts:86]: Not sure we ever discussed a plan for pulling deprecated properties.  I suppose now would be as good a time as any - would that be your recommendation @joeskeen?  And the compiler was throwing an odd warning for the @deprecated tab, it also wasn't rendering very well in our template (just plain text).  So for all of the above I switched to what you see here.

File: projects/cashmere/src/lib/form-field/hc-form-field.component.spec.ts (hunk 8 above)
Comment (andrew-frueh) [reply projects/cashmere/src/lib/form-field/hc-form-field.component.spec.ts:77]: I wasn't sure if some of the empty or commented unit tests were stubs that people thought should be in but never got to.  Happy to remove the commented or `xit` items if you think that's best.

File: projects/cashmere/src/lib/sort/sort.spec.ts (hunk 9 above)
Comment (andrew-frueh) [reply projects/cashmere/src/lib/sort/sort.spec.ts:45]: I think this entire file of unit tests is commented out.  Not sure the history of that, but I'm in favor of removing.  Will do, and same for table.

File: projects/cashmere/src/lib/toaster/hc-toast-ref.ts (hunk 12 above)
Comment (andrew-frueh) [reply projects/cashmere/src/lib/toaster/hc-toast-ref.ts:7]: Wow, great catch.  This has been in the library for ages and never got noticed, but you're absolutely right.  Change made.

File: projects/cashmere/src/lib/form-field/hc-form-field.component.spec.ts (hunk 8 above)
Comment (andrew-frueh) [reply projects/cashmere/src/lib/form-field/hc-form-field.component.spec.ts:77]: I'll go ahead and remove all the commented tests, but I'll leave the `xit` ones assuming they are stubs.

**Comment by andrew-frueh** [comment]
alrighty @joeskeen - think I got all those suggestions addressed.  Let me know if you notice anything else
